### PR TITLE
feat(calendar): add calendar widget

### PR DESCRIPTION
**Pull request by sophacles**
A calendar widget that with some display options and styling for individual dates.

This widget displays the month provided to it in the constructor. (see docs and the example for more info).

One thing to note: This pulls in a new dependency, the [time](https://crates.io/crates/time) crate. Since this adds a dependency, it might be worth putting the widget behind a feature flag - I'm happy to do so if that's desired.

### CODE REVIEW

**Review by joshka — APPROVED** [review]
Looks great! Some minor comments only

File: src/widgets/calendar.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    /// Fill the calendar slots for days not in the current month also, this causes each line to be
+    /// completely filled. If there is an event style for a date, this style will be patched with
+    /// the event's style
+    pub fn show_surrounding(mut self, style: Style) -> Self {
+        self.show_surrounding = Some(style);
+        self
+    }
+
+    /// Display a header containing weekday abbreviations
+    pub fn show_weekdays(mut self, style: Style) -> Self {
```
[the remark this answers is not in the document]
Comment (sophacles) [reply src/widgets/calendar.rs]: Good idea. Updated.

File: examples/calendar.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    // earth day
+    list.add(
+        Date::from_calendar_date(current_year, Month::April, 22).unwrap(),
+        holiday_style,
+    );
+    // star wars day
+    list.add(
+        Date::from_calendar_date(current_year, Month::May, 4).unwrap(),
+        holiday_style,
+    );
```
[the remark this answers is not in the document]
Comment (sophacles) [reply examples/calendar.rs:151]: Thanks!

File: src/widgets/calendar.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    layout::Rect,
+    style::Style,
+    text::{Span, Spans},
+    widgets::{Block, Widget},
+};
+
+use time::{Date, Duration, OffsetDateTime};
+
+/// Display a month calendar for the month containing `display_date`
+pub struct Calendar<'a, S: DateStyler> {
```
[the remark this answers is not in the document]
Comment (sophacles) [reply src/widgets/calendar.rs:25]: It's a good idea. What I'd really like is to set it up for use like:

```
use ratatui::widgets::calendars::{Month, Week /* and so on ...*/};
```
But that deviates from the current pattern a lot.

Thoughts?

[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/calendar.rs:25]: > `use ratatui::widgets::calendars::{Month, Week /* and so on ...*/};`

`widgets::calendars::{Month,Week}` makes 100% sense to me. Great idea.

Btw, the lint error just looks like it's for the commit message (which will go away with a squash).

**Review by joshka — APPROVED** [review]
Changes for header LGTM.

I'm not sure whether the `time` dependency will be an actual problem for anyone (I'd guess it's probably a fundamental enough crate that probably not). The bigger issue on that idea though is that if you do plan to move to using chrono down the line, perhaps feature flagging on whether to use time or chrono calendars could be worth doing up front. I don't have any strong leanings either way on this. It could be best to just leave this until it's an issue.

I'd lean towards adding the change to `calendars::Month` in this PR the name just seems more correct and changing the widget names down the line would break more dependents.

Slapping an approve as no big deal if you don't make that change either.

**Comment by sayanarijit** [comment]
Should we put calendar and other similar new-dependency-introducing widgets behind feature flags? @orhun @mindoodoo

**Comment by sayanarijit** [comment]
> Hmm, what would you name the feature ?

`widget-calendar`?

**Comment by sophacles** [comment]
I put in the machinery for the feature and named it `widget-calendar`. It's easy to change if something else is decided.

**Comment by orhun** [comment]
I didn't understand why you went with `dep:time` instead of `time`, why that is required?

**Comment by sophacles** [comment]
> why is that required?

Apparently it isn't :). I'm still learning about some of these Cargo directives and I think I misunderstood some of what I read.

Pushed a fixup

**Comment by orhun** [comment]
Just tested and it works well. GJ! 🎉

One question though, is it possible to add tests for this?

**Comment by sophacles** [comment]
I can add tests (i think). I'll see what I can get in place over the weekend.

**Comment by sophacles** [comment]
I've added some tests for the widget, as well as for the date styler.

One thing to note, I'm not sure if CI will actually run the tests, because I don't know if it builds with default features or does something "smart". I added a feature `all-widgets` in my latest commit, which currently just specs `widget-calendar`, but as we add other widgets behind feature flags, I presume it will get more entries. It also allows for a simpler testing with `cargo test --features all-widgets`

**Comment by orhun** [comment]
I think you can update `Makefile.toml` like this for running tests for `all-widgets` feature:

```diff
diff --git a/Makefile.toml b/Makefile.toml
index 8afba0a..db9354f 100644
--- a/Makefile.toml
+++ b/Makefile.toml
@@ -100,11 +100,11 @@ args = [
 ]
 
 [tasks.test-crossterm]
-env = { TUI_FEATURES = "serde,crossterm" }
+env = { TUI_FEATURES = "serde,crossterm,all-widgets" }
 run_task = "test"
 
 [tasks.test-termion]
-env = { TUI_FEATURES = "serde,termion" }
+env = { TUI_FEATURES = "serde,termion,all-widgets" }
 run_task = "test"
 
 [tasks.test]
```

**Comment by sophacles** [comment]
Thanks @orhun - that looks like it did the trick, I can calendar tests being run now.

File: src/widgets/calendar.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sophacles) [reply src/widgets/calendar.rs]: Linter errors were taken care of by squashing, thanks for the suggestions!

**Review by orhun — APPROVED** [review]

**Review by joshka — APPROVED** [review]

**Review by mindoodoo — APPROVED** [review]
This looks awesome man !

**Pull request merged**